File: scripts/dnq/minmax.py
import random
from dnq import DNQ, dnq

class MinMax(DNQ):

    def __init__(self, lst: list[int]):
        self.lst = lst

    def is_basic(self):
        return len(self.lst) <= 3
    
    def base_fun(self):
        n = len(self.lst)
        minimum = self.lst[0]
        maximum = self.lst[0]
        for i in range(1, n):
            minimum = min(minimum, self.lst[i])
            maximum = max(maximum, self.lst[i])
        return minimum, maximum
    
    def decompose(self):
        m = int(len(self.lst) / 2)
        a = MinMax(self.lst[:m])
        b = MinMax(self.lst[m:])
        return a, b

    def recombine(self, results):
        minimum = min(results[0][0], results[1][0])
        maximum = max(results[0][1], results[1][1])
        return minimum, maximum


if __name__ == '__main__':
    
    lst = [i for i in range(0, 10)]
    random.shuffle(lst)
    problem = MinMax(lst)
    
    print(lst)
    res = dnq(problem)
    print(res)

File: demos/DivideAndConquer/MinMax.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DivideAndConquer
{
	internal class MinMax<T>(T[] items) : IDivideAndConquerable<(T, T)> where T : IComparable<T>
	{
		public bool IsBasic => items.Length <= 3;

		public (T, T) BaseFun()
		{
			return (items.Min()!, items.Max()!);
		}

		public IDivideAndConquerable<(T, T)>[] Decompose()
		{
			var m = items.Length / 2;
			return new MinMax<T>[] { new(items[..m]), new(items[m..]) };
		}

		public (T, T) Recombine((T, T)[] results)
		{
			return (results.Min(it => it.Item1)!, results.Max(it => it.Item2)!);
		}
	}
}
